File: src/main/java/nieboczek/lifestolen/Formatting.java
package nieboczek.lifestolen;

import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;

public final class Formatting {
    private Formatting() {
    }

    public static MutableComponent toComponent(String text) {
        return Component.literal(text);
    }

    public static MutableComponent red(String text) {
        return toComponent(text).withColor(0xFF3636);
    }

    public static MutableComponent green(String text) {
        return toComponent(text).withColor(0x00FF00);
    }

    public static MutableComponent darkGray(String text) {
        return toComponent(text).withColor(0x404040);
    }

    public static MutableComponent purple(String text) {
        return toComponent(text).withColor(0xA842ED);
    }

    public static MutableComponent niceBlue(String text) {
        return toComponent(text).withColor(0xBBAAE0);
    }

    public static MutableComponent gradientText(String text, int start, int end) {
        MutableComponent result = Component.empty();
        int length = text.length();

        for (int i = 0; i < length; i++) {
            float ratio = (float) i / (length - 1);
            int r = (int) (((start >> 16) & 0xFF) * (1 - ratio) + ((end >> 16) & 0xFF) * ratio);
            int g = (int) (((start >> 8) & 0xFF) * (1 - ratio) + ((end >> 8) & 0xFF) * ratio);
            int b = (int) ((start & 0xFF) * (1 - ratio) + (end & 0xFF) * ratio);

            int rgb = (r << 16) | (g << 8) | b;
            result.append(Component.literal(String.valueOf(text.charAt(i))).withColor(rgb));
        }
        return result;
    }

    public static final class Options {
        public final int color;
        public final boolean bold;
        public final boolean italic;
        public final boolean underline;

        public Options(int color, boolean bold, boolean italic, boolean underline) {
            this.color = color;
            this.bold = bold;
            this.italic = italic;
            this.underline = underline;
        }

        public MutableComponent apply(String text) {
            return Component.literal(text).withColor(color);
        }

        public MutableComponent apply(Component component) {
            return component.copy().withColor(color);
        }
    }
}

File: src/main/kotlin/nieboczek/lifestolen/Formatting.kt
package nieboczek.lifestolen

import net.minecraft.network.chat.Component
import net.minecraft.network.chat.MutableComponent

object Formatting {
    private fun toComponent(text: String): MutableComponent {
        return Component.literal(text)
    }

    fun red(text: String): MutableComponent {
        return toComponent(text).withColor(0xFF3636)
    }

    fun green(text: String): MutableComponent {
        return toComponent(text).withColor(0x00FF00)
    }

    fun darkGray(text: String): MutableComponent {
        return toComponent(text).withColor(0x404040)
    }

    fun purple(text: String): MutableComponent {
        return toComponent(text).withColor(0xA842ED)
    }

    fun niceBlue(text: String): MutableComponent {
        return toComponent(text).withColor(0xBBAAE0)
    }

    fun gradientText(text: String, start: Int, end: Int): MutableComponent {
        val result = Component.empty()
        val length = text.length

        for (i in 0 until length) {
            val ratio = i.toFloat() / (length - i)
            val r = (((start shr 16) and 0xFF) * (1 - ratio) + ((end shr 16) and 0xFF) * ratio).toInt()
            val g = (((start shr 8) and 0xFF) * (1 - ratio) + ((end shr 8) and 0xFF) * ratio).toInt()
            val b = ((start and 0xFF) * (1 - ratio) + (end and 0xFF) * ratio).toInt()

            val rgb = (r shl 16) or (g shl 8) or b
            result.append(Component.literal(text[i].toString()).withColor(rgb))
        }
        return result
    }
}
